protowhat/checks/check_bash_history.py

```diff
@@ -0,0 +1,181 @@
+            )
+            return history if full_history else history[old_history_length:]
+    except FileNotFoundError:
+        return []
+
+
+"""
+Design considerations
+
+use case:
+bash history used multiple times in SCT
+solution:
+separate bash history info update
+
+use case:
+bash history is not empty from earlier exercise or session (that might not use this SCT)
+solution:
+update info in PEC or exercise init
+
+If updating bash history info in PEC, only use bash history in check_correct
+as it only has access to the latest commands
+"""
+
+
+def has_command(state, pattern, msg, fixed=False, commands=None):
+    """Test whether the bash history has a command matching the pattern
+
+    Args:
+        state: State instance describing student and solution code. Can be omitted if used with Ex().
+        pattern: text that command must contain (can be a regex pattern or a simple string)
+        msg: feedback message if no matching command is found
+        fixed: whether to match text exactly, rather than using regular expressions
+        commands: the bash history commands to check against,
+            by default all since the last bash history info update,
+            otherwise the result of calling the helper function
+            ``get_bash_history(full_history=False, bash_history_path=None)``
+            with other arguments can be passed.
+
+    Note:
+        The helper function ``update_bash_history_info(bash_history_path=None)``
+        needs to be called in the pre-exercise code in exercise types that don't have
+        support for bash history features by default.
+
+    Note:
+        If the bash history info is updated every time code is submitted,
+        it's advised to only use this function as the second part of a check_correct
+        to help students debug the command they haven't correctly run yet.
+        If the bash history info is only updated at the start of an exercise,
+        this can be used everywhere as the (cumulative) commands from all submissions are known.
+
+    :Example:
+
+        The goal of an exercise is to use ``man``.
+
+        In the pre-exercise code, put::
+
+            update_bash_history_info()
+
+        In the SCT, check whether a command with ``man`` was used::
+
+            Ex().has_command("$man\s", "Your command should start with ``man ...``.")
+
+    """
+    if commands is None:
+        commands = get_bash_history()
+    if not commands:
+        state.report("Looking for an executed shell command, we didn't find any.")
+    if not state.is_root:
+        raise InstructorError(
+            "`has_command()` should only be called from the root state, `Ex()`."
+        )
+
+    correct = False
+    for command in commands:
+        # similar to has_code
+        if pattern in command if fixed else re.search(pattern, command):
+            correct = True
+            break
+
+    if not correct:
+        state.report(msg)
+
+    return state
+
+
+"""
```

Comment:
If this block is about future ideas and usages it could be useful, however in its current form most of it is not clear for us.

Reply:
I tried to clarify it further.